### PR TITLE
[SPARK-42049][SQL][FOLLOWUP] Always filter away invalid ordering/partitioning

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a follow-up of https://github.com/apache/spark/pull/37525 . When the project list has aliases, we go to the `projectExpression` branch which filters away invalid partitioning/ordering that reference non-existing attributes in the current plan node. However, this filtering is missing when the project list has no alias, where we directly return the child partitioning/ordering.

This PR fixes it.

### Why are the changes needed?

to make sure we always return valid output partitioning/ordering.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @peter-toth

**Review by dongjoon-hyun — APPROVED** [review]
cc @sunchao , @huaxingao , @viirya , too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/AliasAwareOutputExpression.scala [hunk 1]
```diff
@@ -114,7 +114,11 @@ trait AliasAwareQueryOutputOrdering[T <: QueryPlan[T]]
         }
       }.takeWhile(_.isDefined).flatten.toSeq
     } else {
-      orderingExpressions
+      // Make sure the returned ordering are valid (only reference output attributes of the current
```
Comment (peter-toth) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/AliasAwareOutputExpression.scala:117]: Good catch! Thanks for this follow-up!

But, I'm not sure this part is correct as `sortOrder.children`s should be filtered with `_.references.subsetOf(outputSet)` separately.

E.g. this test (is a bit artifical though) fails now:
```
  test("SPARK-42049: Improve AliasAwareOutputExpression - no alias but still prune expressions 2") {
    withSQLConf(SQLConf.OPTIMIZER_EXCLUDED_RULES.key ->
      Seq(CollapseProject.ruleName, ColumnPruning.ruleName).mkString(",")) {
      val df = spark.range(2).select($"id" as "a", $"id" as "b").select($"a")
      val outputOrdering = df.queryExecution.optimizedPlan.outputOrdering

      assert(outputOrdering.size == 1)
      assert(outputOrdering.head.child.asInstanceOf[Attribute].name == "a")
      assert(outputOrdering.head.sameOrderExpressions.size == 0)
    }
  }
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/AliasAwareOutputExpression.scala:117]: good point!

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Hmm, the failed test seems a related one.

**Comment by dongjoon-hyun** [comment]
Oh is it still failing?

**Comment by viirya** [comment]
I think it was failing due to latest commit.

**Comment by cloud-fan** [comment]
The failed test checks invalid ordering and I've updated it.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Merged to master/3.4.
Thank you, @cloud-fan , @peter-toth , @viirya .

**Pull request closed without merging**